lib/AST/StmtManual.cpp

```diff
@@ -19,6 +19,95 @@ namespace pasta {
 
 #ifndef PASTA_IN_BOOTSTRAP
 
+// Follow a macro token's parent chain to the end, and returns the final parent.
+// If the macro token has no parent, this returns std::nullopt;
+std::optional<pasta::Macro>
+GetRootSubstitution(const pasta::MacroToken &macro_token) noexcept {
+  auto root = std::optional(macro_token.Parent());
+  while (root && root->Parent()) {
+    root = root->Parent();
+  }
+  return root;
+}
+
+// Returns the highest substitution that covers this Stmt, if any.
+std::optional<pasta::Macro> Stmt::GetCoveringSubstitution() const noexcept {
+  // If the first token in this Stmt did not come from a macro substitution,
+  // then this Stmt is not covered by a substitution
+  const auto begin_macro_token = BeginToken().MacroLocation();
+  if (!begin_macro_token) {
+    return std::nullopt;
+  }
+
+  // If the last token in this Stmt did not come from a macro substitution,
+  // then this Stmt is not covered by a substitution
+  const auto end_macro_token = BeginToken().MacroLocation();
+  if (!end_macro_token) {
+    return std::nullopt;
+  }
+
+  const auto begin_macro_token_root = GetRootSubstitution(*begin_macro_token);
+  const auto end_macro_token_root = GetRootSubstitution(*end_macro_token);
+
+  // If the root of the begin token and end token are the same, then return that
+  // root
+  if (begin_macro_token_root && end_macro_token_root &&
+      *begin_macro_token_root == *end_macro_token_root) {
+    return begin_macro_token_root;
+  }
+
+  // Otherwise, this statement was composed of multiple substitution trees, so
+  // return nothing
+  return std::nullopt;
+}
+
+// Returns true if we can follow the given token's derived locations to a reach
+// a token that has a macro location whose parent is the specified macro.
+inline bool IsTokenDerivedFromMacro(const Token &token, const Macro &macro) {
+  auto derived_token = std::optional(token);
+  while (derived_token) {
+    if (derived_token->MacroLocation() &&
```

Comment:
Maybe assign to local var to avoid repeated computation.

Reply:
Like this?
```c++
auto derived_token = std::optional(token);
do {
  if (derived_token->MacroLocation() &&
      derived_token->MacroLocation()->Parent() &&
      macro == derived_token->MacroLocation()->Parent()) {
    return true;
  }
} while ((derived_token = derived_token->DerivedLocation()));
```

Reply:
More like:
```c++
for (; derived_token; derived_token = derived_tok->DerivedLocation()) {
  auto macro_tok = derived_token->MacroLocation();
  if (!macro_tok) {
    continue;
  }

  auto containing_macro = macro_tok->Parent();
  if (macro == containing_macro.value()) {
    return true;
  }
}
```

Reply:
TL;DR: A local inside the loop to avoid repeated `->MacroLocation()`.